mex.go

```diff
@@ -175,6 +275,7 @@ type messageExchangeSet struct {
 	// maps are mutable, and are protected by the mutex.
 	exchanges        map[uint32]*messageExchange
 	timeoutExchanges map[uint32]struct{}
+	shutdown         bool
```

Comment:
The placement suggests to me that we're protecting `shutdown` with the mutex too, but it doesn't look like we synchronize reads at all.